src/UpgradeRule/AbstractUpgradeRule.php

```diff
@@ -15,6 +17,26 @@
     protected $warningCollector = [];
 
     /**
+     * Called on a code collection prior to upgrade
+     *
+     * @param CollectionInterface $code
+     * @param CodeChangeSet $changeset
+     */
+    public function beforeUpgrade(CollectionInterface $code, CodeChangeSet $changeset)
```

Comment:
Do we actually need the CodeChangeSet here? I'm not sure what people would expect to put in it?